m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -556,21 +556,27 @@ public void run() {
                 }
             }
 
-            state.set(ProcessorState.SHUTDOWN);
-
             LOG.warn("Processor shutting down");
 
+            if (isShutdown.get()) {
+                // If the processor hit an exception, another processor will be created to resume work on the queue,
+                // so leave the queue as is.
+                drainAndFlush();
+            }
             shutdown();
+            state.set(ProcessorState.SHUTDOWN);
 
             LOG.warn("Processor shut down");
         }
 
-        private void shutdown() {
+        private void drainAndFlush() {
             // Drain queue of any remaining metrics submitted prior to shutdown;
             runNoThrow(this::drainQueue);
             // Flush remaining buffers at last (best effort)
             runNoThrow(this::flushBuffered);
+        }
 
+        private void shutdown() {
```

Comment:
Hmm thinking about this--should the countdown latch call go in `drainAndFlush` instead? Looks like we use that to enforce clean shutdown. As it is now I think we'll count down every time the processor has a crash, meaning that we won't wait when we *actually* shutdown .

cc @alexeykudinkin

Reply:
Good point. There is an edge case here:
1. Processor dies
2. Reporter gets shutdown before new processor gets created

In this scenario, the countdown latch will not be decremented. But this should be fine since it's a rare scenario and we'll just end up hitting the timeout when waiting for the latch and move on.

Reply:
Good point! I think the marginally more correct alternative to using a countdown latch would be to switch to a different primitive (not sure which offhand, but having the equivalent of a CompletableFuture per processor would work). I think this is ok as is though as you mentioned.

Reply:
I've replaced the CountdownLatch with a Semaphore as Alexey suggested to correct this.